### PR TITLE
Add transform option

### PR DESCRIPTION
**Pull request by itrelease**
This will allow to transform payload before send it to channels

### CODE REVIEW

**Comment by chrismccord** [comment]
Can you describe the usecase that you have that requires this feature? Thanks!

**Comment by itrelease** [comment]
@chrismccord I need keys in payload object to be in camelCase

**Comment by chrismccord** [comment]
How about we instead modify the channel.onMessage contract to return the payload? This way you could transform the payload on a channel by channel basis and we don't introduce any new apis. The diff would be a single line change:

```diff
- this.onMessage(event, payload, ref)
+ payload = this.onMessage(event, payload, ref)
```

If you wanted this to apply to all messages, you could write your own function to build new channel for you:

```javascript
let newChannel = (socket, topic, params = {}) => {
  let chan = socket.channel(topic, params)
  chan.onMessage = (event, paylaod, ref) => { return camelize(payload) }
  return chan
}
...
let roomChannel = newChannel(socket, "rooms:lobby")
```

Thoughts?

**Comment by itrelease** [comment]
@chrismccord but with this approach we still need to do this for each channel right?

**Comment by itrelease** [comment]
@chrismccord but it doesn't affect payload in each individual event, like:
```
channel.on('[event_name]', payload => {
  // here payload still with original data
});
```
right?

**Comment by chrismccord** [comment]
No, it would allow you to rewrite the payload. By default we would return it unchanged, but what we pass to the channel.on(..) handlers is the rewritten payload

**Comment by itrelease** [comment]
@chrismccord hmm that's odd, I've tried that but still I get original data

**Comment by itrelease** [comment]
@chrismccord sorry, I was passing original data :) yeah your approach works for me

**Comment by chrismccord** [comment]
Great. Please go ahead with the changes!

**Comment by itrelease** [comment]
@chrismccord should default `onMessage` return just payload or the contract have to be object with payload in it like:
```
onMessage(event, payload, ref){
    return { event, payload, ref };
}
```
?

**Comment by chrismccord** [comment]
I think just the payload is fine, unless we want to allow customizing events. I suppose if you want your event names to be customized, you need it as well, but the issue with that is phoenix has some special events that you can't coerce, like phx_error and phx_close, so I'm hesitant to expose transforming the event name. Same with ref, it's not something you'd ever want to transform

File: web/static/js/phoenix.js [hunk 1]
```diff
@@ -397,7 +397,9 @@ export class Channel {
   // Overridable message hook
   //
   // Receives all events for specialized message handling
-  onMessage(event, payload, ref){}
+  onMessage(event, payload, ref){
+    return payload;
+  }
```
Comment (chrismccord) [inline web/static/js/phoenix.js]: single line and kill the semicolon :)

**Comment by itrelease** [comment]
@chrismccord fair enough

**Comment by chrismccord** [comment]
Thanks!

**Comment by chvanikoff** [comment]
Glad I found this, got into a situation where I have to emulate socket.io behaviour and this change will definitely easy my job. Thanks @itrelease !

**Pull request merged**